yt_dlp/extractor/niconico.py

```diff
@@ -328,36 +344,41 @@ def ping():
-            return mobj.groupdict() if mobj else {}
-
-        protocol = 'niconico_dmc'
-        format_id = '-'.join(map(lambda s: remove_start(s['id'], 'archive_'), [video_quality, audio_quality]))
-        vdict = parse_format_id(video_quality['id'])
-        adict = parse_format_id(audio_quality['id'])
-        resolution = try_get(video_quality, lambda x: x['metadata']['resolution'], dict) or {'height': vdict.get('res')}
-        vbr = try_get(video_quality, lambda x: x['metadata']['bitrate'], float)
+    def _extract_format_for_quality(self, video_id, audio_quality, video_quality, dmc_protocol):
+
+        if not audio_quality['isAvailable'] or not video_quality['isAvailable']:
+            return None
+
+        def extract_video_quality(video_quality):
+            # Example: 480p | 0.9M
+            r = re.match(r'^.*\| ([0-9]*\.?[0-9]*[MK])', video_quality)
+            if not r:
+                # Maybe conditionally throw depending on the settings?
+                return 0
+            return parse_filesize(f'{r.group(1)}B') or 0
+
+        format_id = '-'.join(
+            [remove_start(s['id'], 'archive_') for s in (video_quality, audio_quality)] + [dmc_protocol])
+
+        vid_metadata = video_quality.get('metadata') or {}
+        resolution = vid_metadata.get('resolution') or {}
+        vid_quality = vid_metadata.get('bitrate')
+        is_low = 'low' in video_quality['id']
 
         return {
-            'url': '%s:%s/%s/%s' % (protocol, video_id, video_quality['id'], audio_quality['id']),
+            'url': 'niconico_dmc:%s/%s/%s' % (video_id, video_quality['id'], audio_quality['id']),
             'format_id': format_id,
-            'format_note': 'DMC %s' % try_get(video_quality, lambda x: x['metadata']['label'], compat_str),
+            'format_note': 'DMC ' + vid_metadata['label'] + ' ' + dmc_protocol.upper(),
             'ext': 'mp4',  # Session API are used in HTML5, which always serves mp4
-            'vcodec': vdict.get('codec'),
-            'acodec': adict.get('codec'),
-            'vbr': float_or_none(vbr, 1000) or float_or_none(vdict.get('br')),
-            'abr': float_or_none(audio_quality.get('bitrate'), 1000) or float_or_none(adict.get('br')),
-            'height': int_or_none(resolution.get('height', vdict.get('res'))),
-            'width': int_or_none(resolution.get('width')),
-            'quality': -2 if 'low' in format_id else -1,  # Default quality value is -1
-            'protocol': protocol,
+            'acodec': 'aac',
+            'vcodec': 'h264',
+            'abr': float_or_none(audio_quality['metadata'].get('bitrate'), 1000),
+            'vbr': float_or_none(vid_quality if vid_quality > 0 else extract_video_quality(vid_metadata.get('label')), 1000),
+            'height': resolution.get('height'),
+            'width': resolution.get('width'),
+            'quality': -2 if is_low else None,
+            'protocol': 'niconico_dmc',
+            'expected_protocol': dmc_protocol,
```

Comment:
what's `expected_protocol`?

Reply:
Now there's no way to identify which protocol itself is, since all video now have HTTP and HLS. So it needs to be provided externally to which protocol it should change.